crates/shared/src/current_block.rs

```diff
@@ -114,12 +113,55 @@ where
             .context("failed to get current block number")?
             .as_u64())
     }
+
+    /// get latest `length` blocks
+    /// if successful, function guarantees `length` blocks in Result (does not return partial results)
+    /// `last_block` block is at the end of the resulting vector
+    async fn current_blocks(&self, last_block: u64, length: u64) -> Result<Vec<BlockNumberHash>> {
+        ensure!(
+            length > 0 && last_block > 0 && last_block >= length,
+            "invalid input"
+        );
+
+        let include_txs = helpers::serialize(&false);
+        let mut batch_request = Vec::with_capacity(length as usize);
+        for i in (0..length).rev() {
+            let num =
+                helpers::serialize(&BlockNumber::Number((last_block.saturating_sub(i)).into()));
+            let request = self
+                .transport()
+                .prepare("eth_getBlockByNumber", vec![num, include_txs.clone()]);
+            batch_request.push(request);
+        }
+
+        // send_batch guarantees the size and order of the responses to match the requests
+        let mut batch_response = self
+            .transport()
+            .send_batch(batch_request.iter().cloned())
+            .await?
+            .into_iter();
+
+        batch_request
+            .into_iter()
+            .map(|_req| match batch_response.next().unwrap() {
+                Ok(response) => serde_json::from_value::<web3::types::Block<H256>>(response)
+                    .context("unexpected response format")
+                    .and_then(|response| match response.number {
+                        Some(number) => {
+                            Ok((number.as_u64(), response.hash.context("missing hash")?))
+                        }
+                        None => Err(anyhow!("missing block number")),
+                    }),
```

Comment:
ubernit: We can use context twice right:
```suggestion
                    .and_then(|response| Ok((
                        response.number.context("missing block number")?.to_u64(),
                        response.hash.context("missing hash")?,
                    ))),
```

Reply:
Thanks